plugins/include/amxmisc.inc

```diff
@@ -390,109 +283,239 @@ stock show_activity_id(idtarget, idadmin, const name[], const fmt[], any:...)
 			}
 		}
-		case 1: // hide name from all users
+	}
+	else
+	{
+		if(can_see_admin_name(id) != -1)
 		{
-			while (i++ < MaxClients)
-			{
-				if (is_user_connected(i))
-				{
-					LookupLangKey(keyfmt, charsmax(keyfmt), KeyWithoutName, i);
-
-					// skip the "adminname" argument if not showing name
-					vformat(buffer, charsmax(buffer), keyfmt, 4);
-					client_print(i, print_chat, "%s", buffer);
-				}
-			}
+			replace_activity_data(id, name, buffer, charsmax(buffer));
+			send_activity_message(id, buffer);
 		}
 	}
 }
 
+/**
+ * Sends an amx_show_activity based message that obeys the colored chat
+ * message rules set by the set_activity_rules() function.
+ *
+ * @param id        Client index to display message to
+ * @param fmt       Formatting rules
+ * @param ...       Variable number of formatting parameters
+ *
+ * @noreturn
+ */
+stock send_activity_message(id, const fmt[], any:...)
+{
+	new buffer[192];
+	vformat(buffer, charsmax(buffer), fmt, 3)
+
+	if(activity_color)
+		client_print_color(id, activity_sender, buffer);
+	else
+		client_print(id, print_chat, buffer);
+}
+
+/**
+ * Checks whether or not the client can see the admin name according to the amx_show_activity cvar.
+ *
+ * @note The valid values of amx_show_activity are:
+ *       0 = don't display the message at all
+ *       1 = don't show the admin name to anyone
+ *       2 = show the admin name to everyone
+ *       3 = show the admin name only to admins, hide it from normal users
+ *       4 = show the admin name only to admins, show nothing to normal users
+ *       5 = hide the admin name only from admins, show nothing to normal users
+ *
+ * @param id        Client index
+ *
+ * @return          1 if the client should see the admin name.
+ *                  0 if the client shouldn't see the admin name.
+ *                  -1 if the client shouldn't see the message at all.
+ */
+stock can_see_admin_name(id)
+{
+	switch(activity_init())
+	{
+		case 0: return -1;
+		case 1: return 0;
+		case 2: return 1;
+		case 3: return is_user_admin(id) ? 1 : 0;
+		case 4: return is_user_admin(id) ? 1 : -1;
+		case 5: return is_user_admin(id) ? 0 : -1;
+	}
+
+	return -1;
+}
+
+/**
+ * Replaces the admin name and client prefix placeholders for show_activity() functions inside a string.
+ *
+ * @note If the client isn't supposed to see the admin name, the name
+ *       will simply get removed by the string (replace with "").
+ *
+ * @param id            Client index
+ * @param name          Admin name that will be replaced
+ * @param buffer        Buffer where the name will be replaced
+ * @param len           Maximum buffer length
+ *
+ * @noreturn
+ */
+stock replace_activity_data(id, const name[], buffer[], len)
+{
+	if(contain(buffer, activity_ph_prefix) != -1)
+	{
+		static prefix[10];
+		get_activity_prefix(id, prefix, charsmax(prefix));
+		replace_all(buffer, len, activity_ph_prefix, prefix);
+	}
+
+	static cansee; cansee = can_see_admin_name(id);
```

Comment:
Why do you use `static` here?

Reply:
Because of my understanding that static should be used in functions that are called each frame. I can see that many people are confused on whether this is true or not - https://forums.alliedmods.net/showthread.php?t=309363

Reply:
Just use `new` everywhere.